### PR TITLE
chore(operations): Drag our gnu builder and environment  together

### PR DESCRIPTION
**Pull request by Hoverbear**
This proverbially drags out gnu builder and our environment together into harmony.

It teaches:
* Our **environment** to construct multiple `BASEIMAGE`s using a `BOOTSTRAPPER` of our choosing.
* Throws the gnu release process into that.
* Does the least amount of changing/abstracting necessary to  get started.

### CODE REVIEW

File: scripts/environment/prepare.sh [hunk 1]
```diff
@@ -1,16 +1,36 @@
...
 
+# Rust
+curl https://sh.rustup.rs -sSf | sh -s -- -y --profile minimal
 rustup default "$(cat rust-toolchain)"
 rustup component add rustfmt
 rustup component add clippy
 rustup target add wasm32-wasi
 
 cd scripts
+# Node toolchain
```
Comment (binarylogic) [inline scripts/environment/prepare.sh:12]:
```suggestion
# Ruby toolchain
```

File: scripts/environment/prepare.sh [hunk 2]
```diff
@@ -1,16 +1,36 @@
 #! /usr/bin/env bash
 set -e -o verbose
 
+# Rust
```
Comment (binarylogic) [inline scripts/environment/prepare.sh:4]:
```suggestion
# Rust toolchain
```

File: Makefile [hunk 3]
```diff
@@ -161,16 +164,45 @@ build-dev: ## Build the project in development mode (Supports `ENVIRONMENT=true`
...
-	$(RUN) build-x86_64-unknown-linux-gnu
+build-x86_64-unknown-linux-gnu: ## Build dynamically linked binary in release mode for the x86_64 architecture. (Docker only)
+	$(CONTAINER_TOOL) buildx build \
+		--platform linux/amd64 \
+		--build-arg BASEIMAGE="centos:7" \
+		--build-arg BOOTSTRAPPER="scripts/environment/bootstraps/centos-7.sh" \
+		--file scripts/environment/Dockerfile \
+		--tag $(ENVIRONMENT_UPSTREAM):x86_64-unknown-linux-gnu \
+		.
+	$(CONTAINER_TOOL) run \
```
Comment (Hoverbear) [inline Makefile:175]: I plan to abstract all this to generic musl/gnu builders next PR in this series, then the PR after will abstract over platforms.

**Comment by Hoverbear** [comment]
Trying to iron our a permission issue...

**Comment by jszwedko** [comment]
Poking at this, but it seems like I don't have the `buildx` subcommand even after enabling experimental features:

```
➜  vector git:(release-migration) docker version                              
Client:
 Version:           19.03.6-ce
 API version:       1.40
 Go version:        go1.13.4
 Git commit:        369ce74
 Built:             Fri May 29 04:01:26 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.6-ce
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.4
  Git commit:       369ce74
  Built:            Fri May 29 04:01:57 2020
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          1.3.2
  GitCommit:        ff48f57fc83a8c44cf4ad5d672424a98ba37ded6
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
➜  vector git:(release-migration) docker buildx 
docker: 'buildx' is not a docker command.
See 'docker --help'
```

**Comment by jszwedko** [comment]
I was able to install it separately.

The docs say it should be included though:

> Docker Buildx is included in Docker 19.03 and is also bundled with the following Docker Desktop releases. Note that you must enable the ‘Experimental features’ option to use Docker Buildx.

https://docs.docker.com/buildx/working-with-buildx/

**Comment by Hoverbear** [comment]
@jszwedko  That's so weird! :( Yeah, so, we use buildx already for this. This is just removing some layers of indirection and using a new image.

**Comment by Hoverbear** [comment]
@jszwedko I don't really like this method that much tbh :( @bruceg also can't use it since he uses podman.

**Comment by binarylogic** [comment]
Are you thinking about scrapping this then? I wouldn't worry about supporting podmad for cross-compiling different archs on Docker, unless I'm misunderstanding?

**Comment by Hoverbear** [comment]
No based on our discussions this seems like the best method. :) I just don't like it!

**Comment by Hoverbear** [comment]
Closing for #3358

**Pull request closed without merging**